### ISSUE TITLE
ThisType for Ember.computed and Ember.observer

### ISSUE THREAD

**Issue by maerten** [Microsoft/TypeScript#16699]
**TypeScript Version:**  2.3.4

Thanks for your work on this great project :)

**Context**
_We are experimenting with TypeScript 2.x in Ember, and the new TypeScript 2.x features really enable us to type most of our Ember code. There is however one thing we came across that doesn't seem like something we can solve in our interfaces/declarations:_

It seems that there currently is no way to get the `this` context inside the Ember.computed and Ember.observer pattern.

**`ThisType` not compatible with Ember.computed and Ember.observer?**

We have set up some experimental interfaces for Ember.Object.extend() (inspired by Vue.js' type declarations, @ember/types and various other sources) 
https://github.com/draios/ember-typescript2/blob/master/app/types/ember/ember-types.d.ts#L78-L83 
These types/interfaces are working so far; all methods and hooks have the proper `this` type, and class properties are accessible through getters and setters.

However when it comes to computed properties, there seems to be no way to get the proper `this` context inside the function passed to Ember.computed() (see example below).

**Code**
```ts
export default Ember.Object.extend({
  myProp: true,

  myMethod: function() {
     // myProp is accessible because of ThisType
     this.get('someProperty');
  }
  computedProperty: Ember.computed('someProperty', function() {
     // myProp is not accessible, because `this` is set to type <any>
     let myVal = this.get('someProperty');

     // ... both `this` and `myVal` are of type <any>
     // and no way to fix this from an external interface?
  });
});


```

My question(s); is this Ember pattern supported somehow in TypeScript?

Thanks!

---

N.B. I tried something like the following, but it doesn't work (X becomes of type `{}`).
```
// the function passed to Ember.computed
type ComputedPropertyFunc<T, X> = (this: X) => T

// Ember.computed with one observed property and a function
function computed<T, X>(observedProperty1: string, fn: ComputedPropertyFunc<T, X>): T & ThisType<X>
```

**Comment by DanielRosenwasser** [comment]
This is going to be harder than anticipated. Ember.computed returns a ComputedProperty instance, but it appears that the `get` method specially handles computed properties.

**Comment by dwickern** [comment]
@DanielRosenwasser That's true, although I think it's a separate issue. I was not able to define an `Ember.get` which goes from `(T | ComputedProperty<T>) => T`. There's also [Ember.getProperties](https://emberjs.com/api/classes/Ember.html#method_getProperties) which can have a mixture of values and computed properties.

As a workaround, instead of
`Ember.computed(...): ComputedProperty<T>`

we can declare
`Ember.computed(...): T & ComputedPropertyMethods`

It means that using `obj.someCP` instead of `obj.get('someCP')` will type-check when it really shouldn't.

**Comment by DanielRosenwasser** [comment]
I tried having two separate type parameters, kind of like in Vue's `.d.ts` files: one for data/methods, and the other for computers. I figured with some new inference capabilities in TypeScript 2.4, we could potentially get things working. Unfortunately, the language doesn't really know how to decompose plain objects into intersections all that well at this point. 

That's why it's not a separate issue - because giving that correct information (or something close to it) can help the types flow correctly.

In the meantime, here's a a question that might provide a workaround: is it acceptable to have `Ember.computed` restrict `this.get` to its dependencies? You could write something like:

```ts
computed<K extends string, T>(observed1: K, fn: (this: Record<K, any>) => T): ComputedProperty<T>
computed<K extends string, T>(observed1: K, observed2: K, fn: (this: Record<K, any>) => T): ComputedProperty<T>
computed<K extends string, T>(observed1: K, observed2: K, observed3: K, fn: (this: Record<K, any>) => T): ComputedProperty<T>
```

**Comment by dwickern** [comment]
It would be interesting to restrict `this.get` to the CP dependencies-- you could statically ensure that the dependencies are all listed and typed correctly. Unfortunately this ends up being a not-so-useful subset of Ember API. In practice, computed properties depend on complex paths like `'user.name'`, `'users.[]'`, `'users.@each.name'`, `'users.@each.{name,email}'`. You might also want to call other methods on `this`, but that's not so common in my experience.

`this.get` returning type `any` negates many benefits of having these type declarations in the first place. But getting the actual return type brings us back to the original problem.

---

Ember also offers computed property "macros". They generally accept some key into object where they are declared and transform the value somehow. It would be great if we could infer the context here too, for example:
```js
// non-working example
declare function readOnly<This, K extends keyof This>(this: ThisType<This>, key: K): ComputedProperty[This[K]]

export default Ember.Object.extend({
  str: 'hello',
  strReadonly: Ember.computed.readOnly('str') // type: ComputedProperty<string>
});
```

**Comment by maerten** [comment]
Just saw the release notes on TS2.4, great stuff!


One question about one of the new features:
https://blogs.msdn.microsoft.com/typescript/2017/06/27/announcing-typescript-2-4/
Re 
> Return types as inference targets

Would return type inference work for ThisType<T> too? E.g. if Ember.computed() returns something like `ComputedProperty<string> & ThisType<T>`, and we assign to an object that has a ThisType<T> set?

Maybe i misunderstand this feature, but perhaps thats what @DanielRosenwasser meant in earlier:
> I figured with some new inference capabilities in TypeScript 2.4, we could potentially get things working

**Comment by DanielRosenwasser** [comment]
Hey @maerten, that's exactly what I had in mind. My understanding is that `ThisType` isn't used in these scenarios, but we could still use `this` parameter types in some fashion. For example:

```
computed<T>(func: (this: This) => T): ComputedProperty<T, This>;
```

> In practice, computed properties depend on complex paths like `'user.name'`, `'users.[]'`, `'users.@each.name'`, `'users.@each.{name,email}'`

@dwickern that's a good point, and APIs like that come up quite a bit, but I think getting the basic scenario working may be helpful to a decent number of users.
